crates/ironrdp-rdpdr-native/src/nix/simple.rs

```diff
@@ -0,0 +1,790 @@
+use ironrdp_pdu::PduResult;
+use ironrdp_svc::{impl_as_any, SvcMessage};
+use libc::ftruncate;
+use nix::libc::{closedir, opendir, readdir, DIR};
+
+use ironrdp_rdpdr::pdu::efs::*;
+use ironrdp_rdpdr::pdu::esc::{ScardCall, ScardIoCtlCode};
+use ironrdp_rdpdr::pdu::RdpdrPdu;
+use ironrdp_rdpdr::RdpdrBackend;
+use std::ffi::{CStr, CString};
+use std::io::Read;
+use std::io::{Seek, SeekFrom, Write};
+use std::os::fd::{AsFd, AsRawFd};
+use std::os::unix::fs::MetadataExt;
+
+#[derive(Debug)]
+pub struct SimpleRdpdrBackend {
+    pub file_id: u32,
+    pub file_base: String,
+    pub file_map: std::collections::HashMap<u32, std::fs::File>,
+    pub file_path_map: std::collections::HashMap<u32, String>,
+    pub file_dir_map: std::collections::HashMap<u32, DirBox>,
+}
+
+#[derive(Debug)]
+pub struct DirBox {
+    pub dir: *mut DIR,
+}
+// SAFETY: it is safe to send *DIR to a different thread, and it is only accessed by one thread at one time.
```

Comment:
issue: It’s easy to break this assumption with the current API, the `dir` field is `pub` and the user can do whatever he wants with it:

```rust
let dir1 = DirBox { /* … */ }
let dir2 = DirBox { ptr: dir1.ptr) }
```

At this point, if you send `dir2` to a different thread, we are breaking the thread-safety. I suggest to not leave the field public and encapsulate it better.